File: r/competition_1d_demo.R
# competition_1d_demo.R — One-dimensional candidate competition demo
#
# HOW TO RUN IN RSTUDIO:
#   Session > Set Working Directory > To Source File Location
#   (picks up r/.Renviron which sets SCS_LIB_PATH / DYLD_LIBRARY_PATH)
#   Then source or step through with Cmd+Enter.
devtools::load_all("r/")

library(socialchoicelab)
library(htmlwidgets)

# ---------------------------------------------------------------------------
# Voters — 1D Gaussian cloud generated directly with R's RNG.
# draw_voters() only supports 2D, so we use rnorm() here.
# IC load: 20 voters × 1 seat × 201 frames = 4020 intervals.
# ---------------------------------------------------------------------------
set.seed(123L)
voters_1d <- runif(101L, min = -1, max = 1)   # flat vector of 20 ideal points

cat("voters (1D):", length(voters_1d), "ideal points\n")
cat("range:", range(voters_1d), "\n")

# ---------------------------------------------------------------------------
# Competition — hunter-sticker vs hunter-sticker on the real line.
# Each candidate hunts for votes until it wins a seat, then freezes.
# ---------------------------------------------------------------------------
sm <- stream_manager(42L, c("competition_adaptation_hunter"))

trace <- competition_run(
  competitor_positions = c(-.8, .7),    # 1D: two scalars
  competitor_headings  = c(1.0, -1.0),    # pointing toward centre
  strategy_kinds       = c("hunter_sticker", "hunter_sticker"),
  voter_ideals         = voters_1d,
  dist_config          = make_dist_config(n_dims = 1L),
  engine_config        = make_competition_engine_config(
    seat_count  = 1L,
    seat_rule   = "plurality_top_k",
    max_rounds  = 200L,
    step_config = make_competition_step_config(kind = "fixed", fixed_step_size = 0.06)
  ),
  stream_manager = sm
)
cat("n_rounds:", trace$dims()$n_rounds, "\n")
cat("final positions:\n")
print(trace$final_positions())

# ---------------------------------------------------------------------------
# Overlays — 1D centroid and marginal median (computed inline from voters)
# For 1D traces, pass any named list to select which overlays to show.
# The values are ignored; the function computes mean/median internally.
# Note: pareto_set and geometric_mean are not supported for 1D traces.
# ---------------------------------------------------------------------------

# ── Overlay toggles ───────────────────────────────────────────────────────────
# COMPUTE_IC    — per-voter indifference curves (2 vertical lines per voter).
#                 Load: 101 voters × 1 seat × 201 frames = ~20 k intervals.
# COMPUTE_WINSET — 1D WinSet interval (lo, hi) for each seat holder per frame.
#                 Load: 1 seat × 201 frames = 201 intervals.
COMPUTE_IC     <- TRUE
COMPUTE_WINSET <- TRUE
# ─────────────────────────────────────────────────────────────────────────────

w_canvas <- animate_competition_canvas(
  trace,
  voters           = voters_1d,
  competitor_names = c("Candidate A", "Candidate B"),
  title            = "1D Hunter-Sticker Competition",
  trail            = "fade",
  dim_names        = c("Policy Dimension"),
  overlays         = list(centroid = TRUE, marginal_median = TRUE),
  width            = 400L,
  height           = 400L,
  compute_ic       = COMPUTE_IC,
  compute_winset   = COMPUTE_WINSET
)

# R Viewer
w_canvas

# Browser
# htmlwidgets::saveWidget(w_canvas, "/tmp/competition_1d_demo_r.html", selfcontained = TRUE)
# browseURL("/tmp/competition_1d_demo_r.html")

File: python/competition_1d_demo.py
"""competition_1d_demo.py — One-dimensional candidate competition demo
======================================================================

Parallel to r/competition_1d_demo.R. Generates voters from a 1D Gaussian
distribution, runs a hunter-sticker-vs-hunter-sticker race on the real line,
and opens the canvas animation. Each candidate hunts for votes until it wins
a seat, then freezes in place.

Run from the project root:
  export DYLD_LIBRARY_PATH=$(pwd)/build
  python python/competition_1d_demo.py
"""

from __future__ import annotations

import subprocess

import numpy as np

import socialchoicelab as scl
import socialchoicelab.plots as sclp

# ---------------------------------------------------------------------------
# Voters — 1D Gaussian cloud generated directly.
# draw_voters() only supports 2D, so we use numpy's RNG here.
# IC load: 20 voters × 1 seat × 201 frames = 4020 intervals.
# ---------------------------------------------------------------------------
rng       = np.random.default_rng(123)
voters_1d = rng.normal(0.0, 0.5, (20, 1))  # shape (20, 1) — required by competition_run

print(f"voters (1D): {voters_1d.shape[0]} ideal points")
print(f"range: [{voters_1d.min():.3f}, {voters_1d.max():.3f}]")

# ---------------------------------------------------------------------------
# Competition — hunter-sticker vs hunter-sticker on the real line.
# Each candidate hunts until it wins a seat, then freezes.
# ---------------------------------------------------------------------------
sm = scl.StreamManager(
    master_seed=42,
    stream_names=["competition_adaptation_hunter"],
)

headings = np.array([[1.0], [-1.0]])  # pointing toward centre

with scl.competition_run(
    np.array([[-1.2], [1.2]]),          # 2 competitors, 1 dimension
    ["hunter_sticker", "hunter_sticker"],
    voters_1d,
    competitor_headings=headings,
    dist_config=scl.DistanceConfig(salience_weights=[1.0]),
    engine_config=scl.CompetitionEngineConfig(
        seat_count=1,
        seat_rule="plurality_top_k",
        max_rounds=200,
        step_config=scl.CompetitionStepConfig(kind="fixed", fixed_step_size=0.12),
    ),
    stream_manager=sm,
) as trace:
    n_rounds, n_comp, n_dims = trace.dims()
    print(f"n_rounds={n_rounds}, n_competitors={n_comp}, n_dims={n_dims}")
    print("final positions:")
    print(trace.final_positions())

    # ---------------------------------------------------------------------------
    # Overlays — 1D centroid and marginal median (computed inline from voters)
    # For 1D traces pass any dict to select which overlays to show; values are
    # ignored and the function computes mean/median internally from voters.
    # Note: pareto_set and geometric_mean are not supported for 1D traces.
    # ---------------------------------------------------------------------------

    # ── Overlay toggles ───────────────────────────────────────────────────────
    # COMPUTE_IC     — per-voter indifference curves (2 vertical lines/voter).
    #                  Load: 20 voters × 1 seat × 201 frames = ~4 k intervals.
    # COMPUTE_WINSET — 1D WinSet interval (lo, hi) for each seat holder/frame.
    #                  Load: 1 seat × 201 frames = 201 intervals.
    COMPUTE_IC     = True
    COMPUTE_WINSET = True
    # ─────────────────────────────────────────────────────────────────────────

    canvas_path = "/tmp/competition_1d_demo_py.html"
    sclp.animate_competition_canvas(
        trace,
        voters=voters_1d,
        competitor_names=["Candidate A", "Candidate B"],
        title="1D Hunter-Sticker Competition",
        trail="fade",
        dim_names=("Policy Dimension",),
        overlays={"centroid": True, "marginal_median": True},
        width=900,
        height=400,
        path=canvas_path,
        compute_ic=COMPUTE_IC,
        compute_winset=COMPUTE_WINSET,
    )
    print(f"Canvas saved to {canvas_path}")
    subprocess.Popen(["open", canvas_path])
